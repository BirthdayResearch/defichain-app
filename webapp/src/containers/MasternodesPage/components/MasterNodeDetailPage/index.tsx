import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import isEmpty from 'lodash/isEmpty';
import { Helmet } from 'react-helmet';
import classnames from 'classnames';
import { I18n } from 'react-redux-i18n';
import {
  Button,
  ButtonGroup,
  Row,
  Col,
  FormGroup,
  Input,
  Label,
} from 'reactstrap';
import {
  MdArrowBack,
  MdDelete,
  MdCheckCircle,
  MdErrorOutline,
} from 'react-icons/md';
import { NavLink, RouteComponentProps, Redirect } from 'react-router-dom';
import KeyValueLi from '../../../../components/KeyValueLi';
import { MASTER_NODES_PATH } from '../../../../constants';
import { MasterNodeObject } from '../../masterNodeInterface';
import { resignMasterNode } from '../../reducer';
import styles from '../../masternode.module.scss';
import Header from '../../../HeaderComponent';
import { getPageTitle, getCountdownValue } from '../../../../utils/utility';
import { MasterNodesPageStates } from '../..';
import ViewOnChain from '../../../../components/ViewOnChain';
import { openMasternodeUpdateRestartModal } from '../../../PopOver/reducer';
interface RouteProps {
  hash: string;
}

interface MasterNodeDetailPageProps extends RouteComponentProps<RouteProps> {
  masternodes: MasterNodeObject[];
  isMasterNodeResigning: boolean;
  resignedMasterNodeData: string;
  isErrorResigningMasterNode: string;
  resignMasterNode: (masterNodeHash: string) => void;
  openMasternodeUpdateRestartModal: ({
    isOpen: boolean,
    masternode: MasterNodeObject,
  }) => void;
}

const MasterNodeDetailPage: React.FunctionComponent<MasterNodeDetailPageProps> = (
  props: MasterNodeDetailPageProps
) => {
  const {
    match,
    masternodes = [],
    resignMasterNode,
    isMasterNodeResigning,
    resignedMasterNodeData,
    isErrorResigningMasterNode,
    openMasternodeUpdateRestartModal,
  } = props;
  const hashValue = match.params.hash;
  const masternode: any = masternodes.find((ele: any) => {
    return ele.hash && ele.hash.toString() === hashValue;
  });
  const {
    ownerAuthAddress,
    operatorAuthAddress,
    creationHeight,
    state,
    resignHeight,
    resignTx,
    banHeight,
    banTx,
    hash,
    mintedBlocks,
    isMyMasternode,
  } = masternode || {};

  if (isEmpty(masternode)) {
    return <Redirect to={MASTER_NODES_PATH} />;
  }
  const [
    isConfirmationModalOpen,
    setIsConfirmationModalOpen,
  ] = useState<string>(MasterNodesPageStates.default);
  const [wait, setWait] = useState<number>(5);
  const [allowCalls, setAllowCalls] = useState(false);

  useEffect(() => {
    if (allowCalls) {
      if (
        resignedMasterNodeData &&
        !isMasterNodeResigning &&
        !isErrorResigningMasterNode
      ) {
        setIsConfirmationModalOpen(MasterNodesPageStates.success);
      }
      if (
        !resignedMasterNodeData &&
        !isMasterNodeResigning &&
        isErrorResigningMasterNode
      ) {
        setIsConfirmationModalOpen(MasterNodesPageStates.failure);
      }
    }
  }, [
    isMasterNodeResigning,
    resignedMasterNodeData,
    isErrorResigningMasterNode,
    allowCalls,
  ]);

  useEffect(() => {
    let waitToSendInterval;
    if (isConfirmationModalOpen === MasterNodesPageStates.confirm) {
      let counter = 5;
      waitToSendInterval = setInterval(() => {
        counter -= 1;
        setWait(counter);
        if (counter === 0) {
          clearInterval(waitToSendInterval);
        }
      }, 1000);
    }
    return () => {
      clearInterval(waitToSendInterval);
    };
  }, [isConfirmationModalOpen]);

  useEffect(() => {
    if (!getCountdownValue()) {
      setWait(0);
    }
  });

  const remapMasternodeState = (state: string) => {
    const masternodesLabel = 'containers.masterNodes.masterNodesPage';
    switch (state) {
      case 'ENABLED':
        return I18n.t(`${masternodesLabel}.created`);
      case 'RESIGNED':
        return I18n.t(`${masternodesLabel}.resigned`);
      case 'PRE_RESIGNED':
        return I18n.t(`${masternodesLabel}.preresigned`);
      case 'PRE_ENABLED':
        return I18n.t(`${masternodesLabel}.preenabled`);
      case 'PRE_BANNED':
        return I18n.t(`${masternodesLabel}.prebanned`);
      case 'BANNED':
        return I18n.t(`${masternodesLabel}.banned`);
      default:
        return state;
    }
  };

  return (
    <div className='main-wrapper'>
      <Helmet>
        <title>
          {getPageTitle(
            I18n.t(
              'containers.masterNodes.masternodeDetailPage.masternodeDetailTitle',
              {
                hash,
              }
            )
          )}
        </title>
      </Helmet>
      <Header>
        <Button
          to={MASTER_NODES_PATH}
          tag={NavLink}
          color='link'
          className='header-bar-back'
        >
          <MdArrowBack />
          <span className='d-lg-inline'>
            {I18n.t('containers.masterNodes.masternodeDetailPage.masternode')}
          </span>
        </Button>
        <h1>
          {I18n.t(
            'containers.masterNodes.masternodeDetailPage.masternodeDetail'
          )}
          &nbsp;
        </h1>
        {isMyMasternode && (
          <div className='d-flex align-items-center'>
            <FormGroup check>
              <Label check className='switch'>
                <Input
                  type='checkbox'
                  checked={masternode.isEnabled}
                  onChange={() => {
                    openMasternodeUpdateRestartModal({
                      isOpen: true,
                      masternode: {
                        ...masternode,
                        isEnabled: !masternode.isEnabled,
                      },
                    });
                  }}
                />
                &nbsp;
                {I18n.t(`containers.masterNodes.masterNodesPage.enable`)}
              </Label>
            </FormGroup>
            <ButtonGroup className='ml-1'>
              <Button
                color='link'
                onClick={() =>
                  setIsConfirmationModalOpen(MasterNodesPageStates.confirm)
                }
              >
                <MdDelete />
                <span>
                  {I18n.t(
                    'containers.masterNodes.masternodeDetailPage.resignMasterNode'
                  )}
                </span>
              </Button>
            </ButtonGroup>
          </div>
        )}
      </Header>
      <div className='content'>
        <section className='mb-5'>
          <KeyValueLi
            label={I18n.t('containers.masterNodes.masternodeDetailPage.state')}
            value={remapMasternodeState(state)}
          />
          <KeyValueLi
            label={I18n.t(
              'containers.masterNodes.masternodeDetailPage.registered'
            )}
            value={`${creationHeight}`}
          />
          <KeyValueLi
            label={I18n.t(
              'containers.masterNodes.masternodeDetailPage.mintedBlocks'
            )}
            value={`${mintedBlocks}`}
          />
          <KeyValueLi
            label={I18n.t(
              'containers.masterNodes.masternodeDetailPage.resignHeight'
            )}
            value={`${resignHeight}`}
          />
          <KeyValueLi
            label={I18n.t(
              'containers.masterNodes.masternodeDetailPage.banHeight'
            )}
            value={`${banHeight}`}
          />
          <KeyValueLi
            label={I18n.t(
              'containers.masterNodes.masternodeDetailPage.ownerAddress'
            )}
            value={ownerAuthAddress}
            copyable={true}
            uid='address'
          />
          <KeyValueLi
            label={I18n.t(
              'containers.masterNodes.masternodeDetailPage.operatorAddress'
            )}
            value={operatorAuthAddress}
            copyable={true}
            uid='address'
          />
          <KeyValueLi
            label={I18n.t(
              'containers.masterNodes.masternodeDetailPage.resignTx'
            )}
            copyable={true}
            value={resignTx}
          />

          <KeyValueLi
            label={I18n.t('containers.masterNodes.masternodeDetailPage.banTx')}
            copyable={true}
            value={banTx}
          />
          <KeyValueLi
            label={I18n.t('containers.masterNodes.masternodeDetailPage.hash')}
            copyable={true}
            value={hash}
          />
        </section>
      </div>
      <footer
        className={classnames({
          'footer-bar': true,
          'd-none': isConfirmationModalOpen === MasterNodesPageStates.default,
        })}
      >
        <div
          className={classnames({
            'd-none': isConfirmationModalOpen !== MasterNodesPageStates.confirm,
          })}
        >
          <div className='footer-sheet'>
            <dl className='row'>
              <dd className='col-12'>
                <span className='h2 mb-0'>
                  {I18n.t(
                    'containers.masterNodes.masternodeDetailPage.confirmation'
                  )}
                </span>
              </dd>
            </dl>
          </div>
          <Row className='justify-content-between align-items-center'>
            <Col className='d-flex justify-content-end'>
              <Button
                color='link'
                className='mr-3'
                onClick={() =>
                  setIsConfirmationModalOpen(MasterNodesPageStates.default)
                }
              >
                {I18n.t(
                  'containers.masterNodes.masternodeDetailPage.noButtonText'
                )}
              </Button>
              <Button
                color='primary'
                onClick={() => {
                  setAllowCalls(true);
                  resignMasterNode(hash);
                }}
                disabled={wait > 0 ? true : false}
              >
                {I18n.t(
                  'containers.masterNodes.masternodeDetailPage.yesButtonText'
                )}
                &nbsp;
                <span className='timer'>{wait > 0 ? wait : ''}</span>
              </Button>
            </Col>
          </Row>
        </div>
        <div
          className={classnames({
            'd-none': isConfirmationModalOpen !== MasterNodesPageStates.success,
          })}
        >
          <div className='footer-sheet'>
            <div className='text-center'>
              <MdCheckCircle className='footer-sheet-icon' />
              <p>
                {`${I18n.t(
                  'containers.masterNodes.masternodeDetailPage.successText'
                )}`}
              </p>
              <p>{resignedMasterNodeData}</p>
            </div>
          </div>
          <div className='d-flex align-items-center justify-content-center'>
            <ViewOnChain txid={resignedMasterNodeData} />
            <Button color='primary' to={MASTER_NODES_PATH} tag={NavLink}>
              {I18n.t(
                'containers.masterNodes.masternodeDetailPage.backToMasternodePage'
              )}
            </Button>
          </div>
        </div>
        <div
          className={classnames({
            'd-none': isConfirmationModalOpen !== MasterNodesPageStates.failure,
          })}
        >
          <div className='footer-sheet'>
            <div className='text-center'>
              <MdErrorOutline
                className={classnames({
                  'footer-sheet-icon': true,
                  [styles[`error-dialog`]]: true,
                })}
              />
              <p>{isErrorResigningMasterNode}</p>
            </div>
          </div>
          <div className='d-flex align-items-center justify-content-center'>
            <Button color='primary' to={MASTER_NODES_PATH} tag={NavLink}>
              {I18n.t(
                'containers.masterNodes.masternodeDetailPage.backToMasternodePage'
              )}
            </Button>
          </div>
        </div>
      </footer>
    </div>
  );
};

const mapStateToProps = (state) => {
  const {
    masterNodes: {
      masternodes,
      isMasterNodeResigning,
      resignedMasterNodeData,
      isErrorResigningMasterNode,
    },
  } = state;
  return {
    masternodes,
    isMasterNodeResigning,
    resignedMasterNodeData,
    isErrorResigningMasterNode,
  };
};

const mapDispatchToProps = {
  resignMasterNode: (masterNodeHash: string) =>
    resignMasterNode({ masterNodeHash }),
  openMasternodeUpdateRestartModal,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MasterNodeDetailPage);
