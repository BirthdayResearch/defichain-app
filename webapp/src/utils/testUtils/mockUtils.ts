import PersistentStore from '../persistentStore';
import * as isElectron from '../isElectron';
import * as Notifications from '../notifications';
import log from 'loglevel';
import { runSaga } from 'redux-saga';
import axios from 'axios';

const mockPersistentStore = (getData, setData) => {
  PersistentStore.get = jest.fn(() => getData);
  PersistentStore.set = jest.fn(() => setData);
  return PersistentStore;
};

const mockAxios = mockFunc => {
  axios.create = jest.fn().mockImplementation(() => ({
    post: mockFunc,
  }));
};
const dispatchedFunc = async (method, payload = {}) => {
  const dispatched = Array();
  let action = {};
  if (payload) action = { payload };
  await runSaga(
    {
      dispatch: action => dispatched.push(action),
    },
    method,
    action
  );
  return dispatched;
};

const mockisElectron = () => {
  return jest.spyOn(isElectron, 'isElectron');
};
const mockIpcRenderer = () => {
  return jest.spyOn(isElectron, 'ipcRendererFunc');
};

const spyLogger = () => {
  return jest.spyOn(log, 'error');
};

const mockNotification = () => {
  return (
    jest
      .spyOn(Notifications, 'default')
      // tslint:disable-next-line: no-empty
      .mockImplementation(() => {})
  );
};
export {
  mockPersistentStore,
  mockAxios,
  dispatchedFunc,
  mockisElectron,
  mockIpcRenderer,
  spyLogger,
  mockNotification,
};
