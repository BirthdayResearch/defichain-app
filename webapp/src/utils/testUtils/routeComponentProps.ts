import { createMemoryHistory, createLocation } from 'history';
import { match } from 'react-router';
const history = createMemoryHistory();
const path = `/route/:id`;
const match: match<{ id: string }> = {
  isExact: false,
  path,
  url: path.replace(':id', '1'),
  params: { id: '1' },
};

const location = createLocation(match.url);
delete history.location.key;
delete location.key;

export { history, match, location };
