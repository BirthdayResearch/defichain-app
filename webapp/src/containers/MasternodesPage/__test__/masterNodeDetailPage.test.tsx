import React from 'react';
import MasterNodeDetailPage from '../components/MasterNodeDetailPage';
import { match } from 'react-router';
import { Provider } from 'react-redux';
import store from '../../../app/rootStore';
import { shallow } from 'enzyme';
import { history } from '../../../utils/testUtils/routeComponentProps';
import { createLocation } from 'history';
const path = `/route/:hash`;
const match: match<{ hash: string }> = {
  isExact: false,
  path,
  url: path.replace(':hash', '1'),
  params: { hash: '1' },
};
const location = createLocation(match.url);
delete location.key;
const updatedLocation = Object.assign({}, location);
const updatedHistory = Object.assign({}, history, {
  location: updatedLocation,
});

describe('Master Node Detail Page', () => {
  it('should check for snapshot', () => {
    const props = {
      masternodes: [],
      resignMasterNode: () => {},
      isMasterNodeResigning: true,
      resignedMasterNodeData: 'resignedMasterNodeData',
      isErrorResigningMasterNode: 'isErrorResigningMasterNode',
    };
    const wrapper = shallow(
      <Provider store={store}>
        <MasterNodeDetailPage
          match={match}
          {...props}
          history={updatedHistory}
          location={updatedLocation}
        />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
