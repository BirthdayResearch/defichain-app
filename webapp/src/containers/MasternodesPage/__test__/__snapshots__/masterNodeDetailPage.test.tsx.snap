// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Master Node Detail Page should check for snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(MasterNodeDetailPage)
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/route/1",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    isErrorResigningMasterNode="isErrorResigningMasterNode"
    isMasterNodeResigning={true}
    location={
      Object {
        "hash": "",
        "pathname": "/route/1",
        "search": "",
        "state": undefined,
      }
    }
    masternodes={Array []}
    match={
      Object {
        "isExact": false,
        "params": Object {
          "hash": "1",
        },
        "path": "/route/:hash",
        "url": "/route/1",
      }
    }
    resignMasterNode={[Function]}
    resignedMasterNodeData="resignedMasterNodeData"
  />
</ContextProvider>
`;
